### DISCUSSION TITLE
How to write board games with ebiten properly?

### DISCUSSION THREAD

**Discussion by ghost** [hajimehoshi/ebiten#1782] [Q&A] [answered]
I'm making a board game(reversi) with ebiten, which let humans against computer.
Because it is a turn-based game, so i decided to execute the rounding in a separate goroutine, like:
```
...
func (g *game) round() {
    for {
        // black side move
        black.move()
        // white side move
        white.move()
    }
}

func main() {
    ...
    go game.round()
    if err := ebiten.RunGame(game); err != nil {
        panic(err)
    }
}
...
``` 
I chose to create a new goroutine rather than put it into ```func (g *game) Update() error``` is because my AI took a century to solve a single move (~10 secs). If I put it into  ```func (g *game) Update() error```, it will cause the screen freezes because it spent most of the time to run the AI part.
Creating a goroutine in the ```func (g *game) Update() error``` doesn't works, neither. Since Update was calling every frame, it will spawn thousands of goroutines doing the same thing.

My solution above (```go game.round()```) works fine when I ran it natively.
But when running under wasm, the screen stuck when the AI is running (even I've created a goroutine besides frontend)
(maybe wasm doesn't support multitasking?)

Is there any other better way to dealing with it?
Thanks.

**Comment by hajimehoshi** [comment, marked as answer]
The concurrent switching might not work well on Wasm. What about calling `runtime.Gosched()`?

In general, you should avoid a busy loop that never sleeps. I'd insert sleeping `time.Sleep(time.Millisecond)`, or, adjust the amount of the calculation by channels.

**Reply by ghost** [reply]
I omitted the time.Sleep() in the code to make it simpler, actually it does have a sleep of 50 milliseconds. (sorry)
My AI was implemented in minimax algorithm (recursive function).
I just tried to insert runtime.Gosched() in minimax, but this way seems not working.
Inserting sleep (1 millisecond) in minimax does make the screen active, but the AI literally stopped to move (maybe it never switch back to AI) (I also tried microsecond and nanosecond)

**Reply by hajimehoshi** [reply]
So probably there are some busy logics that never returns for a while. I'd add more sleeping for a lot of suspicious locations.

**Reply by ghost** [reply]
OK, I'll find out where's the problem in my code.
I'm wonder that will wasm have a better concurrency support in the future
Thanks for answering the question !

**Reply by Zyko0** [reply]
Hey @lemon37564 if you want to avoid goroutines or Gosched and stuff
You might want to tweak your AI so that it computes only a part of the search each tick (each call to `Update`), like only a few nodes by tick, you keep that state to keep going the next tick.
Whenever you covered enough, you should have a result ready to update your game's logic and you can clear/update your AI's state for a new turn after the player played its turn

edit: should have written this outside this thread sorry
